### DISCUSSION TITLE
Amending into a Mapping inside a class

### DISCUSSION THREAD

**Discussion by fhalde** [apple/pkl#652] [Q&A] [answered]
I have the following class

```
class User {
  name: String
  properties: Mapping<String, String>?
}
```

and the following JSON which is a partial representation of User
```
{ "name": "foo" }
```

I read this json in my pkl and use the [deepToTyped](https://github.com/apple/pkl-pantry/blob/main/packages/pkl.experimental.deepToTyped/deepToTyped.pkl) to convert it into a User instance

```
me: User = deepToTyped.apply(User, read(<json>))
```

How can someone amend the properties Mapping?

```
me: User = (deepToTyped.apply(User, read(<json>))) {
 properties = new Mapping {
   ["x"] = "ok"
 }
}
```

works, however, it wipes out the properties if they were there in JSON
```
{ "name": "foo", "properties": { "y": "notok" } }
```

**Comment by fhalde** [comment]
My requirements is to ensure, both the remote properties & locally defined properties get merged into the final properties Mapping

**Comment by fhalde** [comment]
At first I thought I could just set a default in the class
```
properties: Mapping<String, String>? = new Mapping {}
```

but this failed with an error
```
Expected value of type `Mapping`, but got type `Dynamic`.
Value: new Dynamic { ["x"] = ? }
```

```
me: User = (deepToTyped.apply(User, read(<json>))) {
 properties {
   ["x"] = "ok"
 }
}
```

**Comment by fhalde** [comment]
This worked after I did the following change

```
properties: Mapping<String, String>? = new Mapping { }
```

to
```
properties: Mapping<String, String> = new Mapping { }
```

removing the optionality `?`
not sure why though

**Comment by bioball** [comment, marked as answer]
I haven't dug into this, but, it's probably because of these lines: https://github.com/apple/pkl-pantry/blob/b7d8b68eb980d3cb6f6babb16537580425419fd6/packages/pkl.experimental.deepToTyped/deepToTyped.pkl#L129-L132

`properties: Mapping<String, String>?` is sugar for: `properties: Mapping<String, String>? = Null(new Mapping<String, String> {})`

`Null(new Mapping<String, String> {})` is a _different_ null than the `null` literal; they behave differently when amended.

This is probably a bug in `deepToTyped`, feel free to create a GitHub issue for it (or submit a PR to fix it)

**Reply by fhalde** [reply]
you're right !!
